### PR TITLE
Refactor interpreted-data interface in preparation for other interpretations than SARIF

### PR DESCRIPTION
**Pull request by hvitved**
Having this change in makes it easier to keep https://github.com/github/vscode-codeql/pull/705 up to date.

### CODE REVIEW

File: extensions/ql-vscode/src/pure/interface-types.ts [hunk 1]
```diff
@@ -44,7 +45,14 @@ export interface PreviousExecution {
...
 
-export interface Interpretation {
+export type SarifInterpretationData = {
+  t: 'SarifInterpretationData';
+} & sarif.Log;
+
+// Add more interpretation data kinds when needed (e.g., graph data)
+export type InterpretationData = SarifInterpretationData;
+
+export interface InterpretationT<T> {
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/pure/interface-types.ts:55]:
```suggestion
export interface Interpretation<T> {
```

File: extensions/ql-vscode/src/pure/interface-types.ts [hunk 2]
```diff
@@ -53,9 +61,11 @@ export interface Interpretation {
    * they appear in the sarif file.
    */
   sortState?: InterpretedResultsSortState;
-  sarif: sarif.Log;
+  data: T;
 }
 
+export type Interpretation = InterpretationT<InterpretationData>;
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/pure/interface-types.ts:67]:
```suggestion
export type Interpretation = Interpretation<InterpretationData>;
```

File: extensions/ql-vscode/src/view/result-tables.tsx [hunk 3]
```diff
@@ -345,8 +350,15 @@ class ResultTable extends React.Component<ResultTableProps, {}> {
     switch (resultSet.t) {
       case 'RawResultSet': return <RawTable
         {...this.props} resultSet={resultSet} />;
-      case 'SarifResultSet': return <PathTable
-        {...this.props} resultSet={resultSet} />;
+      case 'InterpretedResultSet': {
+        const data = resultSet.interpretation.data;
+        switch (data.t) {
+          case 'SarifInterpretationData': {
+            const sarifResultSet = { ...resultSet, interpretation: { ...resultSet.interpretation, data } };
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/view/result-tables.tsx:357]: This is confusing to me. It doesn't look like anything is changing here. Could you add a comment to explain what you're doing?

File: extensions/ql-vscode/src/pure/interface-types.ts [hunk 4]
```diff
@@ -10,13 +10,14 @@ export const SELECT_TABLE_NAME = '#select';
...
 
 export type RawTableResultSet = { t: 'RawResultSet' } & RawResultSet;
-export type PathTableResultSet = {
-  t: 'SarifResultSet';
+export type InterpretedResultSet<T> = {
+  t: 'InterpretedResultSet';
   readonly schema: ResultSetSchema;
   name: string;
-} & Interpretation;
+  interpretation: InterpretationT<T>;
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/pure/interface-types.ts:17]:
```suggestion
  interpretation: Interpretation<T>;
```

File: extensions/ql-vscode/src/pure/interface-types.ts [hunk 5]
```diff
@@ -44,7 +45,14 @@ export interface PreviousExecution {
   durationSeconds: number;
 }
 
-export interface Interpretation {
+export type SarifInterpretationData = {
+  t: 'SarifInterpretationData';
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/pure/interface-types.ts:49]: Here, you are added a nested `t` field. This seems to complicate things somewhat. Would it make sense to keep only one `t` field at the top level, and specify `RawResultSet`, `SarifResultSet`, and (eventually) `GraphResultSet`?

I don't remember your other PR, so maybe there's a good reason to keep this.

File: extensions/ql-vscode/src/pure/interface-types.ts [hunk 6]
```diff
@@ -53,9 +61,11 @@ export interface Interpretation {
    * they appear in the sarif file.
    */
   sortState?: InterpretedResultsSortState;
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/pure/interface-types.ts:63]: Does this field make sense for graph results?

File: extensions/ql-vscode/src/pure/interface-types.ts (hunk 2 above)
Comment (hvitved) [reply extensions/ql-vscode/src/pure/interface-types.ts:67]: This doesn't appear to work, I guess you can't have both a generic type and a non-generic type with the same name?

File: extensions/ql-vscode/src/view/result-tables.tsx (hunk 3 above)
Comment (hvitved) [reply extensions/ql-vscode/src/view/result-tables.tsx:357]: Right, this dispatch on `t` only makes sense when we have other interpretation types, so I can move it in to the other PR.

File: extensions/ql-vscode/src/pure/interface-types.ts (hunk 6 above)
Comment (hvitved) [reply extensions/ql-vscode/src/pure/interface-types.ts:63]: No, so I'll move it into `SarifInterpretationData`.

File: extensions/ql-vscode/src/pure/interface-types.ts (hunk 5 above)
Comment (hvitved) [reply extensions/ql-vscode/src/pure/interface-types.ts:49]: This is in preparation for
```ts
export type GraphInterpretationData = {
  t: 'GraphInterpretationData';
  dot: string[];
};
```
(from [the other PR](https://github.com/github/vscode-codeql/pull/705)) so we can do dispatch on the types (this is how to do tagged unions in Typescript, right?)

**Comment by hvitved** [comment]
Rebased to resolve merge conflicts

**Review by Alitld — APPROVED** [review]

**Comment by aeisenberg** [comment]
Superseded by https://github.com/github/vscode-codeql/pull/1111

**Pull request closed without merging**